specs/core/0_beacon-chain.md

```diff
@@ -1886,135 +1949,301 @@ def process_transfer(state: BeaconState, transfer: Transfer) -> None:
+    ]
+```
 
-* Let `crosslink_data_root` be `state.latest_crosslinks[shard].crosslink_data_root`
-* Let `attesting_validator_indices(crosslink_committee, crosslink_data_root)` be the union of the [validator](#dfn-validator) index sets given by `[get_attestation_participants(state, a.data, a.aggregation_bitfield) for a in current_epoch_attestations + previous_epoch_attestations if a.data.shard == shard and a.data.crosslink_data_root == crosslink_data_root]`.
-* Let `winning_root(crosslink_committee)` be equal to the value of `crosslink_data_root` such that `get_total_balance(state, attesting_validator_indices(crosslink_committee, crosslink_data_root))` is maximized (ties broken by favoring lexicographically smallest `crosslink_data_root`).
-* Let `attesting_validators(crosslink_committee)` be equal to `attesting_validator_indices(crosslink_committee, winning_root(crosslink_committee))` for convenience.
-* Let `total_attesting_balance(crosslink_committee) = get_total_balance(state, attesting_validators(crosslink_committee))`.
+```python
+def get_previous_epoch_matching_head_attestations(state: BeaconState) -> List[PendingAttestation]:
+    return [
+        a for a in state.previous_epoch_attestations if
+        a.data.beacon_block_root == get_block_root(state, a.data.slot)
+    ]
+```
 
-Define the following helpers to process attestation inclusion rewards and inclusion distance reward/penalty. For every attestation `a` in `previous_epoch_attestations`:
+**Note**: Total balances computed for the previous epoch might be marginally different than the actual total balances during the previous epoch transition. Due to the tight bound on validator churn each epoch and small per-epoch rewards/penalties, the potential balance difference is very low and only marginally affects consensus safety.
 
-* Let `inclusion_slot(state, index) = a.inclusion_slot` for the attestation `a` where `index` is in `get_attestation_participants(state, a.data, a.aggregation_bitfield)`. If multiple attestations are applicable, the attestation with lowest `inclusion_slot` is considered.
-* Let `inclusion_distance(state, index) = a.inclusion_slot - a.data.slot` where `a` is the above attestation.
+```python
+def get_winning_root_and_participants(state: BeaconState, shard: Shard) -> Tuple[Bytes32, List[ValidatorIndex]]:
+    all_attestations = state.current_epoch_attestations + state.previous_epoch_attestations
+    valid_attestations = [
+        a for a in all_attestations if a.data.latest_crosslink == state.latest_crosslinks[shard]
+    ]
+    all_roots = [a.data.crosslink_data_root for a in valid_attestations]
+    
+    def get_attestations_for(root) -> List[PendingAttestation]:
+        return [a for a in valid_attestations if a.data.crosslink_data_root == root]
+        
+    winning_root = max(all_roots, key=lambda r: get_attesting_balance(state, get_attestations_for(r)))
```

Comment:
1. It was noted that, in `get_winning_root`: ties broken by favoring lexicographically smallest `crosslink_data_root`.
2. In Python, [max() and min() on a list will return the first in the list that matches the max()/min() in the event of a tie](https://stackoverflow.com/a/6783051).
3. It seems now the tie break is depending on the order in `state.current_epoch_attestations + state.previous_epoch_attestations`? This change seems worth to be noted in release note.

/cc @NIC619